contracts/contracts/zapper/WOETHCCIPZapper.sol

```diff
@@ -0,0 +1,143 @@
+// SPDX-License-Identifier: MIT
+pragma solidity ^0.8.0;
+
+import { IRouterClient } from "@chainlink/contracts-ccip/src/v0.8/ccip/interfaces/IRouterClient.sol";
+import { Client } from "@chainlink/contracts-ccip/src/v0.8/ccip/libraries/Client.sol";
+import { IERC20 } from "@chainlink/contracts-ccip/src/v0.8/vendor/openzeppelin-solidity/v4.8.0/contracts/token/ERC20/IERC20.sol";
+import { IERC4626 } from "./../../lib/openzeppelin/interfaces/IERC4626.sol";
+import { IOETHZapper } from "./../interfaces/IOETHZapper.sol";
+
+/**
+ * @title WOETH CCIP Zapper Contract
+ * @notice Handles logic to easily zap ETH on mainnet to WOETH on L2s like Arbitrum.
+ * @author Origin Protocol Inc
+ */
+
+contract WOETHCCIPZapper {
+    /**
+     * @dev Event emitted when a zap occurs
+     * @param messageId Unique message identifier for each zap
+     * @param sender Address initiating the zap
+     * @param recipient Recipient address at destination chain
+     * @param amount Amount of ETH zapped
+     */
+    event Zap(
+        bytes32 indexed messageId,
+        address sender,
+        address recipient,
+        uint256 amount
+    );
+
+    /**
+     * @dev The destination chain selector
+     */
+    uint64 public immutable destinationChainSelector;
+
+    /**
+     * @dev The WOETH source chain (Mainnet)
+     */
+    IERC4626 public immutable woethOnSourceChain;
+
+    /**
+     * @dev The WOETH destination chain (Arbitrum)
+     */
+    IERC20 public immutable woethOnDestinationChain;
+
+    /**
+     * @dev The OETH zapper contract address
+     */
+    IOETHZapper public immutable oethZapper;
+
+    /**
+     * @dev The CCIP router contract address
+     */
+    IRouterClient public immutable ccipRouter;
+
+    /**
+     * @dev The OETH token contract address
+     */
+    IERC20 public immutable oeth;
+
+    constructor(
+        address _ccipRouter,
+        uint64 _destinationChainSelector,
+        IERC4626 _woethOnSourceChain,
+        IERC20 _woethOnDestinationChain,
+        IOETHZapper _oethZapper,
+        IERC20 _oeth
+    ) {
+        ccipRouter = IRouterClient(_ccipRouter);
+        destinationChainSelector = _destinationChainSelector;
+        woethOnSourceChain = _woethOnSourceChain;
+        woethOnDestinationChain = _woethOnDestinationChain;
+        oethZapper = _oethZapper;
+        oeth = _oeth;
+
+        // Max allowance for Router and WOETH contracts
+        _oeth.approve(address(_woethOnSourceChain), type(uint256).max); // for wrapping
+        _woethOnSourceChain.approve(address(_ccipRouter), type(uint256).max); // for zapping
+    }
+
+    /**
+     * @notice Accepts ETH, zaps for OETH, wraps it for WOETH and sends it to the destination chain (arbitrum)
+     * @param receiver The address of the EOA on the destination chain
+     * @return messageId The ID of the message that was sent
+     */
+    function zap(address receiver) public payable returns (bytes32 messageId) {
```

Comment:
This can be external, not public?

Reply:
I think `_zap` is called in the receive, so you could mark the `zap` as external